strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,406 @@
+    (if ``threshold=True``) or mean number of photons (if ``threshold=False``) is also required.
+    The initial value ``n_mean`` should be set high since varying :math:`\theta` can only lower
+    the mean number of clicks or photons.
+
+    The mapping from :math:`\theta` to :math:`W(\theta)` is specified using the ``embedding``
+    argument. TODO mention embedding module
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(4, 0.7, seed=1967)
+    >>> A = nx.to_numpy_array(g)
+    >>> embedding = Exp(4)  # TODO does this need updating?
+    >>> vgbs = VGBS(A, 3, embedding, threshold=True)
+    >>> params = np.array([0.05, 0.1, 0.02, 0.01])
+    >>> vgbs.A(params)
+    array([[0.        , 0.30298161, 0.31534653, 0.31692721],
+           [0.30298161, 0.        , 0.30756059, 0.30910225],
+           [0.31534653, 0.30756059, 0.        , 0.32171695],
+           [0.31692721, 0.30910225, 0.32171695, 0.        ]])
+    >>> vgbs.n_mean(params)
+    2.299036355948707
+    >>> vgbs.generate_samples(vgbs.A(params), 2)
+    array([[1, 1, 1, 0],
+           [0, 1, 0, 1]])
+
+    Args:
+        A (array): the input adjacency matrix :math:`A`
+        n_mean (float): the initial mean number of clicks or photons
+        embedding: the method of converting from trainable parameters :math:`\theta` to
+            :math:`W(\theta)` #TODO: any update from other PR?
+        threshold (bool): determines whether to use GBS in threshold or PNR mode
+        samples (array): an optional array of samples from :math:`A` used to speed up gradient
+            calculations #TODO: more info
+    """
+
+    def __init__(
+        self,
+        A: np.ndarray,
+        n_mean: float,
+        embedding,
+        threshold: bool,
+        samples: Optional[np.ndarray] = None,
+    ):
+        if not np.allclose(A, A.T):
+            raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+        self.A_init = rescale_adjacency(A, n_mean, threshold)
+        self.A_init_samples = None
+        self.embedding = embedding
+        self.threshold = threshold
+        self.n_modes = len(A)
+        self.add_A_init_samples(samples)
+
+    def W(self, params: np.ndarray) -> np.ndarray:
+        r"""Calculate the diagonal matrix of weights :math:`W` that depends on the trainable
+        parameters :math:`\theta`.
+
+        **Example usage:**
+
+        >>> vgbs.W(params)
+        array([[0.97530991, 0.        , 0.        , 0.        ],
+               [0.        , 0.95122942, 0.        , 0.        ],
+               [0.        , 0.        , 0.99004983, 0.        ],
+               [0.        , 0.        , 0.        , 0.99501248]])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the diagonal matrix of weights
+        """
+        return np.sqrt(np.diag(self.embedding(params)))
+
+    def A(self, params: np.ndarray) -> np.ndarray:
```

Comment:
(Minor suggestion): wondering here: could the `WAW` name work? :thinking: as later used, could `A(params)` lead to some confusion when seen out of context?

Reply:
That is a very good question! We've discussed this a bit and are so far sticking with the idea that we should not overburden users with more technical namings, and in the end will present it as `A(\theta)`, simply as a matrix that is a function of the params (which is the main thing a user needs to know).